### PR TITLE
accounts/keystore: fix race-condition in Import/ImportECDSA

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
Calling keystore.Import or keystore.ImportECDSA might result in a race.
If they are called in parallel, they both check whether a key is already in the keystore.
The keystore is empty so both threads insert their key into the keystore.
This however results in a keystore that contains a key twice.

I personally do not like the additionally mutex in the structure, but I currently don't see another way to do this.

additional: Tests for exporting/importing keys

closes https://github.com/ethereum/go-ethereum/issues/19575#event-3230244782

### CODE REVIEW

File: accounts/keystore/keystore.go [hunk 1]
```diff
@@ -67,7 +67,8 @@ type KeyStore struct {
 	updateScope event.SubscriptionScope // Subscription scope tracking current live listeners
 	updating    bool                    // Whether the event notification loop is running
 
-	mu sync.RWMutex
+	mu    sync.RWMutex
+	impMu sync.RWMutex // Import Mutex locks the import to prevent two insertions from racing
```
Comment (holiman) [inline accounts/keystore/keystore.go:71]: I'd prefer it written out, spend another two characters for better readability: `importMu`

File: accounts/keystore/keystore.go [hunk 2]
```diff
@@ -67,7 +67,8 @@ type KeyStore struct {
 	updateScope event.SubscriptionScope // Subscription scope tracking current live listeners
 	updating    bool                    // Whether the event notification loop is running
 
-	mu sync.RWMutex
+	mu       sync.RWMutex
+	importMu sync.RWMutex // Import Mutex locks the import to prevent two insertions from racing
```
Comment (holiman) [inline accounts/keystore/keystore.go:71]: Please use a `sync.Mutex` instead, since you're not using the `RLock`

Comment (holiman) [reply accounts/keystore/keystore.go:71]: sorry about late nitpick :)

File: accounts/keystore/keystore.go [hunk 3]
```diff
@@ -67,7 +67,8 @@ type KeyStore struct {
 	updateScope event.SubscriptionScope // Subscription scope tracking current live listeners
 	updating    bool                    // Whether the event notification loop is running
 
-	mu sync.RWMutex
+	mu       sync.RWMutex
```
Comment (MariusVanDerWijden) [inline accounts/keystore/keystore.go:70]: Use sync.Mutex

File: accounts/keystore/keystore.go [hunk 4]
```diff
@@ -443,12 +444,19 @@ func (ks *KeyStore) Import(keyJSON []byte, passphrase, newPassphrase string) (ac
 	if err != nil {
 		return accounts.Account{}, err
 	}
+	ks.importMu.Lock()
+	defer ks.importMu.Unlock()
+	if ks.cache.hasAddress(key.Address) {
+		return accounts.Account{}, fmt.Errorf("account already exists")
```
Comment (karalabe) [inline accounts/keystore/keystore.go:450]: `fmt.Errorf` -> `errors.New`

**Pull request merged**